grpc-common/README.md

```diff
@@ -0,0 +1,195 @@
+    }
+}
+
+service MyService {
+    rpc Get (TestApi.GetRequest) returns (TestApi.GetResponse) {}
+}
+```
+
+The Scala trait:
+```scala
+import io.grpc.Status
+import scala.concurrent.Future
+
+case class MyRequest(names: Seq[String])
+
+case class MyResponse(results: Map[String, Int])
+
+trait ServerTrait {
+  def get(request: MyRequest): Future[Either[Status, MyResponse]]
+}
+```
+
+Map the instance of `ServerTrait` to generated service stub:
+```scala
+import com.avast.cactus.grpc.server._  // this adds the `mappedToService` method
+
+val impl: ServerTrait = ???
+
+val service: io.grpc.ServerServiceDefinition = impl.mappedToService[MyServiceImplBase]()
+```
+
+Build the server in a standard way and add the service to it.
+
+### Server interceptors
+
+You can decorate the handler with two kinds of interceptors:
+1. Asynchronous - instances of `ServerAsyncInterceptor` passed to the mapping function
+1. Synchronous using the `ServerInterceptors.intercept` method (gRPC standard)
+
+Read the [section below](#providing-own-interceptors) before developing your own interceptor.
+
+### Providing context
+
+gRPC has no option how to pass received headers to the handler by default - programmer must handle them in interceptors (and possibly
+passing something into the handler through [Context](https://grpc.io/grpc-java/javadoc/io/grpc/Context.html)). Cactus provides context
+mapping for the handler, e.g.
+```scala
+case class MyRequest(names: Seq[String])
+
+case class MyResponse(results: Map[String, Int])
+
+case class MyContextContent(i: Int, s: String)
+
+case class MyContext(theHeader: String, content: MyContextContent)
+
+trait MyApi {
+  def get(request: MyRequest, ctx: MyContext): Future[Either[Status, MyResponse]]
+}
+```
+
+Note the context class has to
+* be a case class
+* be a second parameter of the API method (name of the parameter is not important).
+
+The `MyContext` instance is created by mapping incoming `Context` in following way:
+```scala
+val ctx = Context.current()
+
+for {
+  theHeader <- Option(ContextKeys.get[String]("theHeader").get(ctx))
+  content <- Option(ContextKeys.get[MyContextContent]("content").get(ctx))
+} yield {
+  MyContext(theHeader, content)
+}
+```
+1. `theHeader` is header received from the client. Headers are automatically passed to the `Context` to be available for this - no action required
+1. `content` is a case class inserted into the `Context` by an interceptor like this one:
+    ```scala
+    new ServerAsyncInterceptor {
+      override def apply(m: GrpcMetadata): Future[Either[Status, GrpcMetadata]] = {
+        val cont = MyContextContent(42, "jenda")
+        
+        Future.successful {
+          Right(m.copy(context = m.context.withValue(ContextKeys.get[MyContextContent]("content"), cont)))
+        }
+      }
+    }
+    ```
+Note that names of `MyContext` class fields have to match keys in `Context`. If some field is missing, an `INVALID_ARGUMENT` status is
```

Comment:
If the context case class has an `Option` parameter and the key is missing, will the `INVALID_ARGUMENT` status be still returned?

Reply:
Good question. Currently the context class mapping does NOT support `Option`s (when mapping headers).  Do you think it should support it?

Reply:
Not necessarily, but would be nice to have. Could be a future improvement.